planner/bounds.go

```diff
@@ -0,0 +1,107 @@
+package planner
+
+import "github.com/influxdata/flux/values"
+
+// EmptyBounds is a time range containing only a single point
+var EmptyBounds = &Bounds{
+	Start: values.Time(0),
+	Stop:  values.Time(0),
+}
+
+// Bounds is a range of time
+type Bounds struct {
+	Start values.Time
+	Stop  values.Time
+}
+
+// BoundedProcedureSpec represents any procedure that
+// introduces time bounds for the data it produces.
+type BoundedProcedureSpec interface {
+	TimeBounds() *Bounds
+}
+
+// ComputeBounds computes the time bounds for a
+// plan node from the bounds of its predecessors.
+func ComputeBounds(node PlanNode) error {
+	var bounds *Bounds
+
+	for _, pred := range node.Predecessors() {
+
+		if pred.Bounds() != nil && bounds == nil {
+			bounds = pred.Bounds()
+		}
+		if pred.Bounds() != nil && bounds != nil {
+			bounds = bounds.Union(pred.Bounds())
+		}
+	}
+
+	s, bdd := node.ProcedureSpec().(BoundedProcedureSpec)
```

Comment:
I'm wondering what the behavior for the flux function `shift` should be here.  It's not bounded in the same way that `range` is, it just translates the time columns in its input by a fixed amount.

Maybe instead of `BoundedProcedureSpec` as an interface we need something like this
```go
type BoundsAwareProcedureSpec interface {
    TimeBounds(predBounds *plan.Bounds) plan.Bounds
}
```
Both `range` and `shift` could implement this, but `range` would intersect, and `shift` would shift.

The behavior you have here is also the behavior of the old planner, so if you don't want to do this now, maybe that's okay.

Reply:
I completely overlooked `shift`. You are right, procedures should implement their own `TimeBounds` method. This is an easy fix. I'll add it.